src/lua/version.lua

```diff
@@ -0,0 +1,248 @@
+-- https://www.tarantool.io/en/doc/latest/release/policy/#versioning-policy
+--
+-- Tarantool's versions use SemVer-like format:
+--     x.y.z-typeN-commit-ghash
+--
+-- * x.y.z - major, minor, patch release numbers;
+-- * typeN - pre-release name and its optional number: alpha1, rc10. Optional;
+-- * commit - number of commits since the latest release. Optional;
+-- * ghash - latest commit hash in format g<hash>. Optional.
+--
+-- Differences with the semver docs:
+--
+-- * Pre-release identifiers are not separated. Not 'alpha.1' but 'alpha1'.
+-- * Pre-release identifiers cannot be nested. Not x.y.z-alpha.beta.
+-- * Pre-release identifier always includes non-numeric type: e.g. beta, rc.
+-- * Build metadata is denoted by appending hyphen, not plus sign. Not
+--   x.y.z-typeN+commit-ghash, but x.y.z-typeN-commit-ghash.
+--
+-- Note, that ghash itself and everything after it (e.g. -dev suffix) is
+-- ignored and doesn't affect comparing of the versions.
+--
+
+local ffi   = require('ffi')
+local utils = require('internal.utils')
+
+local check_param = utils.check_param
+
+--
+-- Cdata is used in order to allow comparison of the version object with
+-- strings. Since relational metamethods do not support mixed types in Lua,
+-- using tables is not possible.
+--
+ffi.cdef([[
+    struct version {
+        /** Major number. */
+        uint32_t major;
+        /** Minor number. */
+        uint32_t minor;
+        /** Patch number. */
+        uint32_t patch;
+        /** Pre-release type: e.g. entrypoint, alpha. */
+        const char *_prerelease_name_pointer;
+        /** Pre-release number: e.g. in alpha1 it's 1. */
+        uint16_t _prerelease_num;
+        /** Number of commits since release. */
+        uint32_t commit;
+    };
+]])
+
+local prerelease_name_weight = {
+    entrypoint = 10,
+    alpha = 20,
+    beta = 30,
+    rc = 40,
+}
+
+local function version_new(major, minor, patch, prerelease, commit)
+    commit = commit or 0
+    check_param(major, 'major', 'number')
+    check_param(minor, 'minor', 'number')
+    check_param(patch, 'patch', 'number')
+    check_param(commit, 'commit', 'number')
+
+    local prerelease_name = ''
+    local prerelease_num = 0
+    if prerelease then
+        check_param(prerelease, 'prerelease', 'string')
+        prerelease_name, prerelease_num = prerelease:match('^(%a+)(%d+)')
+        if not prerelease_name then
+            prerelease_name = prerelease:match('^(%a+)')
+        else
+            prerelease_num = tonumber(prerelease_num)
+        end
+        if not prerelease_name_weight[prerelease_name] then
+            local msg = "Unknown release type '%s'"
+            box.error(box.error.ILLEGAL_PARAMS,
+                      string.format(msg, prerelease_name))
+        end
+    end
+
+    return ffi.new('struct version', {
+        major = major,
+        minor = minor,
+        patch = patch,
+        _prerelease_name_pointer = prerelease_name,
```

Comment:
`prerelease_name` may be garbage-collected while the version object is still in use.

```lua
local ffi = require('ffi')
ffi.cdef([[struct foo { const char *bar; };]])
local foo = ffi.new('struct foo', {string.rep('x', 1024 * 1024)})
collectgarbage('collect')
print(ffi.string(foo.bar))
```

<details><summary>Segmentation fault</summary>

```
Segmentation fault
  code: SEGV_MAPERR
  addr: 0x40bdc020
  context: 0x5a59a16f6340
  siginfo: 0x5a59a16f6470
  rax      0x20               32
  rbx      0x407a953c         1081775420
  rcx      0x100              256
  rdx      0x40bdc020         1086177312
  rsi      0x40016008         1073831944
  rdi      0x40bdc020         1086177312
  rsp      0x74ec18c80be8     128557376867304
  rbp      0x407a9378         1081774968
  r8       0x40db8628         1088128552
  r9       0x100              256
  r10      0x0                0
  r11      0x203              515
  r12      0x0                0
  r13      0x0                0
  r14      0x407aa0a0         1081778336
  r15      0x5a599f5416e5     99340971677413
  rip      0x74ec1bb8b75d     128557426194269
  eflags   0x10283            66179
  cs       0x33               51
  gs       0x0                0
  fs       0x0                0
  cr2      0x40bdc020         1086177312
  err      0x4                4
  oldmask  0x0                0
  trapno   0xe                14
Current time: 1731495081
Please file a bug at https://github.com/tarantool/tarantool/issues
Attempting backtrace... Note: since the server has already crashed,
this may fail as well
#1  0x5a599f483b61 in crash_collect+256
#2  0x5a599f48459d in crash_signal_cb+100
#3  0x74ec1ba45320 in __sigaction+80
#4  0x74ec1bb8b75d in __nss_database_lookup+14205
#5  0x5a599f54196d in lj_cf_ffi_string+648
#6  0x5a599f4f25c7 in lj_BC_FUNCC+70
#7  0x5a599f50010b in lua_pcall+969
#8  0x5a599f45ef03 in luaT_call+45
#9  0x5a599f45046c in lua_main+283
#10 0x5a599f451086 in run_script_f+3073
#11 0x5a599f1a7183 in fiber_cxx_invoke(int (*)(__va_list_tag*), __va_list_tag*)+34
#12 0x5a599f490757 in fiber_loop+219
#13 0x5a599f8c0f18 in coro_init+120
Aborted
```

</details>

Reply:
I didn't manage to reproduce the segfault with version module itself. Tried to run the following script:

<details><summary>Details</summary>
<p>

```lua
local v = require('version')
local a = v('1.1.1-entrypoint1-45')
collectgarbage('collect')
print(a.prerelease)

```

</p>
</details>

However, I anyway changed implementation so that now it doesn't have any strings in cdef structure out of harm's way. It's anyway better to save weights and not names themselves, this way it's easier to compare, the code even looks cleaner, IMHO